jlm/rvsdg/theta.cpp

```diff
@@ -113,9 +136,8 @@ ThetaNode::MapPostLoopVar(const rvsdg::input & result) const
     // There is nothing sensible to return here.
     throw std::logic_error("cannot map loop continuation predicate to loop variable");
   }
-  auto peer = MapOutputToInputIndex(result.index() - 1);
-  return LoopVar{ peer ? input(*peer) : nullptr,
-                  peer ? subregion()->argument(*peer) : nullptr,
+  return LoopVar{ input(result.index()),
+                  subregion()->argument(result.index()),
```

Comment:
These two lines (139, 140) need `result.index() - 1`. This fixed the crash

Reply:
@caleridas We tracked the crash to this

Reply:
That is the site of the crash but not the cause: I restructured DNE such that unbalanced input/output nodes never occur in loop treatment -- for jlm/llvm/opt/DeadNodeElimination.cpp, that is, that also has unit test coverage to validate this invariant.

However I suspect that the actual error is in jlm/hls/backend/rvsdg2rhls/DeadNodeElimination.cpp which has some open-coded treatment of input removal that then also breaks invariants. I will keep adding invariant checks until finding this, but since it does not reproduce on my machine I will have a lot of CI roundtrips to see where it is coming from. Maybe if you have a reproducer you could also take a look?

Reply:
I agree that there are issues in other places, since the local reproduction was non-deterministic.

But do you not agree that the input and argument should have `index - 1`, when `index` is the index of the result? The list of results has the extra predicate input, so all other indices need to be offset by 1? @caleridas

Reply:
oooh sorry my bad -- yes you are spot on, that should do it